linkerd/http/retry/src/peek_trailers.rs

```diff
@@ -94,55 +108,103 @@ impl<B: Body> PeekTrailersBody<B> {
+                data: Some(Err(error)),
+                trailers: None,
+            },
+            // The body yielded a TRAILERS frame. We are done.
+            Some(Ok(Err(trailers))) => Self::Unary {
+                data: None,
+                trailers: Some(Ok(trailers)),
+            },
+            // The body yielded a DATA frame. Check for a second frame, without yielding again.
+            Some(Ok(Ok(first))) => {
+                if let Some(second) = body
+                    .frame()
+                    .map(|f| f.map(|r| r.map(Self::split_frame)))
+                    .now_or_never()
+                {
+                    // The second frame is available. Let's inspect it and determine what to do.
+                    match second {
+                        // The body is finished. There is not a TRAILERS frame.
+                        None => Self::Unary {
+                            data: Some(Ok(first)),
+                            trailers: None,
+                        },
+                        // We immediately yielded a result, but it was an error. Alas!
+                        Some(Err(error)) => Self::Unary {
+                            data: Some(Ok(first)),
+                            trailers: Some(Err(error)),
+                        },
+                        // We immediately yielded another frame, but it was a second DATA frame.
+                        // We hold on to each frame, but we cannot wait for the TRAILERS.
+                        Some(Ok(Ok(second))) => Self::Unknown {
+                            first: Some(Ok(first)),
+                            second: Some(Ok(second)),
+                            inner: body.into_inner(),
+                        },
+                        // The body immediately yielded a second TRAILERS frame. Nice!
+                        Some(Ok(Err(trailers))) => Self::Unary {
+                            data: Some(Ok(first)),
+                            trailers: Some(Ok(trailers)),
+                        },
+                    }
+                } else {
+                    // If we are here, the second frame is not yet available. We cannot be sure
+                    // that a second DATA frame is on the way, and we are no longer willing to
+                    // await additional frames. There are no trailers to peek.
+                    Self::Unknown {
+                        first: None,
+                        second: None,
+                        inner: body.into_inner(),
+                    }
+                }
+            }
         };
 
-        if trailers.is_some() {
+        if body.peek_trailers().is_some() {
             tracing::debug!("Buffered trailers frame");
         }
 
-        Self {
-            inner: body,
-            first_data,
-            trailers,
-        }
+        body
     }
 
-    /// Returns a response with an inert [`PeekTrailersBody<B>`].
+    /// Splits a `Frame<T>` into a `Result<T, E>`.
```

Comment:
It might be slightly clearer as an `Either`? We can borrow a an Either type from `futures::future::Either`.